File: src/main/java/com/dg/watcher/watching/surveying/SizeSurveying.java
package com.dg.watcher.watching.surveying;

import com.dg.watcher.watching.saving.SizeEntry;
import java.util.ArrayList;
import static com.dg.watcher.base.Const.CONVERSION_FACTOR_BYTE_TO_MEGABYTE;
import static com.dg.watcher.watching.surveying.SizeSurveyingResult.SIZE_THRESHOLD_EXCEEDED;
import static com.dg.watcher.watching.surveying.SizeSurveyingResult.SIZE_THRESHOLD_MET;


public class SizeSurveying {
    public static SizeSurveyingResult surveySizes(ArrayList<SizeEntry> entries, float thresholdInMb) {
        if(atLeastTwoSizesRecorded(entries)) {
            float latestSizeDifferenceInMb = calculateLatestSizeDifferenceInMb(entries);

            if(sizeThresholdExceeded(latestSizeDifferenceInMb, thresholdInMb)) {
                return SIZE_THRESHOLD_EXCEEDED;
            }
        }

        return SIZE_THRESHOLD_MET;
    }

    private static boolean atLeastTwoSizesRecorded(ArrayList<SizeEntry> entries) {
        return entries.size() > 1;
    }

    private static float calculateLatestSizeDifferenceInMb(ArrayList<SizeEntry> entries) {
        SizeEntry previousApk = entries.get(entries.size() - 2);
        SizeEntry latestApk = entries.get(entries.size() - 1);

        return (latestApk.getSizeInByte() - previousApk.getSizeInByte())
                / CONVERSION_FACTOR_BYTE_TO_MEGABYTE;
    }

    private static boolean sizeThresholdExceeded(float differenceInMb, float thresholdInMb) {
        return differenceInMb > thresholdInMb;
    }
}

File: src/main/java/com/dg/watcher/watching/surveying/SizeSurveying.kt
package com.dg.watcher.watching.surveying

import com.dg.watcher.base.CONVERSION_FACTOR_BYTE_TO_MEGABYTE
import com.dg.watcher.watching.saving.SizeEntry
import com.dg.watcher.watching.surveying.SizeSurveyingResult.SIZE_THRESHOLD_EXCEEDED
import com.dg.watcher.watching.surveying.SizeSurveyingResult.SIZE_THRESHOLD_MET


fun surveySizes(sizes: List<SizeEntry>, thresholdInMb: Float): SizeSurveyingResult {
    if(atLeastTwoSizesRecorded(sizes)) {
        val latestSizeDifferenceInMb = calculateLatestSizeDifferenceInMb(sizes)

        if(sizeThresholdExceeded(latestSizeDifferenceInMb, thresholdInMb)) {
            return SIZE_THRESHOLD_EXCEEDED
        }
    }

    return SIZE_THRESHOLD_MET
}

private fun atLeastTwoSizesRecorded(sizes: List<SizeEntry>) = sizes.size > 1

private fun calculateLatestSizeDifferenceInMb(sizes: List<SizeEntry>): Float {
    val sizeOfPreviousApk = sizes[sizes.lastIndex - 1].sizeInByte
    val sizeOfLatestApk = sizes[sizes.lastIndex].sizeInByte

    return (sizeOfLatestApk - sizeOfPreviousApk) / CONVERSION_FACTOR_BYTE_TO_MEGABYTE
}

private fun sizeThresholdExceeded(differenceInMb: Float, thresholdInMb: Float) = differenceInMb > thresholdInMb